File: frontend/src/types/accuracyReport.ts
/** Mirrors backend app/schemas/accuracy_report.py - the response shape for
 * GET/POST .../accuracy-report (see CompanyDocumentsHubView's Accuracy panel,
 * which is the only consumer). */
export interface AccuracyMismatch {
  period_label: string;
  field: string;
  expected: number;
  // null when the field wasn't extracted at all, distinct from a wrong value.
  got: number | null;
  source_excerpt: string | null;
  source_page: number | null;
  statement_id: string | null;
}

export interface IdentityCheckResult {
  rule_name: string;
  passed: boolean;
  expected: number;
  actual: number;
  delta: number;
}

export interface AccuracyScorecard {
  fields_compared: number;
  exact_matches: number;
  mismatches: AccuracyMismatch[];
  identity_checks_passed: number;
  identity_checks_total: number;
  identity_check_results: IdentityCheckResult[];
  ground_truth_available: boolean;
  ground_truth_fixture: string | null;
}

export interface AccuracyReport {
  id: string;
  document_id: string;
  pipeline_version: string;
  scorecard: AccuracyScorecard;
  created_at: string;
}

File: backend/app/schemas/accuracy_report.py
import uuid
from datetime import datetime

from pydantic import ConfigDict

from app.schemas.base import AppBaseModel


class AccuracyReportRequest(AppBaseModel):
    document_id: uuid.UUID


class AccuracyMismatch(AppBaseModel):
    period_label: str
    field: str
    expected: float
    # None when the field wasn't extracted at all (present in ground truth,
    # missing from the document's current FinancialStatement rows) - distinct
    # from a wrong-but-present value.
    got: float | None
    source_excerpt: str | None
    source_page: int | None
    statement_id: uuid.UUID | None


class IdentityCheckResult(AppBaseModel):
    rule_name: str
    passed: bool
    expected: float
    actual: float
    delta: float


class AccuracyScorecard(AppBaseModel):
    fields_compared: int
    exact_matches: int
    mismatches: list[AccuracyMismatch]
    identity_checks_passed: int
    identity_checks_total: int
    identity_check_results: list[IdentityCheckResult]
    ground_truth_available: bool
    ground_truth_fixture: str | None


class AccuracyReportRead(AppBaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    document_id: uuid.UUID
    pipeline_version: str
    # AccuracyReport.scorecard is a plain JSONB dict - Pydantic validates it
    # against this nested model automatically (dicts are valid input for a
    # nested BaseModel field regardless of from_attributes on the parent).
    scorecard: AccuracyScorecard
    created_at: datetime
